staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go

```diff
@@ -102,19 +108,36 @@ func WithAudit(handler http.Handler, requestContextMapper request.RequestContext
 		// running requests, this will be the second audit event.
 		defer func() {
 			if r := recover(); r != nil {
+				defer panic(r)
+				ev.Stage = auditinternal.StagePanic
 				ev.ResponseStatus = &metav1.Status{
-					Code: http.StatusInternalServerError,
+					Code:    http.StatusInternalServerError,
+					Reason:  metav1.StatusReasonInternalError,
+					Message: fmt.Sprintf("APIServer panic'd: %v", r),
 				}
 				sink.ProcessEvents(ev)
-				panic(r)
+				return
 			}
 
-			if ev.ResponseStatus == nil {
+			// if no StageResponseStarted event was sent b/c neither a status code nor a body was sent, fake it here
+			if ev.ResponseStatus == nil && longRunningSink != nil {
 				ev.ResponseStatus = &metav1.Status{
-					Code: 200,
+					Code:    http.StatusOK,
+					Reason:  metav1.StatusSuccess,
+					Message: "connection closed early",
 				}
+				ev.Stage = auditinternal.StageResponseStarted
+				longRunningSink.ProcessEvents(ev)
 			}
 
+			ev.Stage = auditinternal.StageResponseComplete
+			if ev.ResponseStatus == nil {
```

Comment:
Move this above the previous if block, and you only need to build the fake status once.